# Business Case - Bronze Layer Exploration

## Import

In [69]:
import pandas as pd

customers = pd.read_csv('../data/customers_in_.csv', sep=';')
orders = pd.read_csv('../data/orders_in_.csv', sep=';')
products = pd.read_csv('../data/products_in_.csv', sep=';')
subscriptions = pd.read_csv('../data/subscriptions_in_.csv', sep=';')

pd.set_option('display.float_format', '{:.0f}'.format)

In [70]:
customers

,client_id,client_country,join_date,segment
0,AP54723-EURO,IT,2019-01-04,Ret-E
1,AP56503-EURO,IT,2019-03-06,Ret-E
2,AP58323-EURO,IT,2019-05-17,Ret-E
3,AP58361-EURO,IT,2019-05-19,Ret-E
4,AP60022-EURO,IT,2019-07-23,Ret-E
...,...,...,...,...
92037,GC50639-EURO,IT,2017-11-14,Ret-E
92038,GC29384-EURO,IT,2012-05-08,Ret-E
92039,EN3414-EURO,IT,2019-09-27,Ret-E
92040,LM28190-EURO,IT,2018-02-15,Ret-E


In [71]:
orders

,client_id,ordercode,renewcode,orderdate,expiredate,subscription_id,product_id,quantity,productprice,currency,duration_months,flag_recurring
0,AO9002-EURO,00000009zrna,00000009zrna,2022-03-11,2023-04-14,23395131,1839,1,0,EUR,12,1
1,AI6077-EURO,558d7f5b91a78,000000036tuu,2018-06-22,2019-06-26,9013343,359,1,26,EUR,12,1
2,GM31287-EURO,000000041ors,00000005mw9a,2020-01-14,2021-01-17,15359823,28181,0,0,EUR,12,1
3,AC84123-EURO,000000046mkc,5e2287feeacd0,2020-01-18,2021-02-07,15507842,57441,1,59,EUR,12,1
4,MG45759-EURO,00000002xlul,00000005ayzb,2019-10-24,2020-10-23,13845917,1588,0,25,EUR,12,1
...,...,...,...,...,...,...,...,...,...,...,...,...
209995,FP34701-EURO,00000005q2w2,00000005q2w2,2020-02-01,2021-02-01,17138679,608,1,0,EUR,12,1
209996,AMN62-EURO,536cdc7372500,5ad80efd5e8c5,2018-04-19,2019-05-09,8039312,1747,0,3,EUR,12,1
209997,NC6090-EURO,588f18376c2c2,5c17283c925f8,2018-12-17,2020-01-06,11489705,28882,0,0,EUR,12,1
209998,MDC1485-EURO,58134b1ac028a,0000000974fx,2021-10-18,2022-10-28,11085983,319,1,35,EUR,12,1


In [72]:
products

,product_id,product_code,product_desc,product_subgroup_desc,product_group_desc,product_subcat_desc,"product_cat_desc,,,,"
0,1106,chgownde,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
1,1107,chgowndj,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
2,1112,chgownec,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
3,1117,chgownes,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
4,1119,chgowneu,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
...,...,...,...,...,...,...,...
1457,56502,deds2up_c_xe3_1240v5,deds2up_c_xe3_1240v5,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"
1458,56503,deds2up_raid_h330,deds2up_raid_h330,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"
1459,56562,domfashion,Domain,Domain Names,Domainnames,Domain Names,"Domains,,,,"
1460,56623,deds2_dellds452,Server Dell DS 4.5.2,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"


In [73]:
subscriptions

,subscription_id,creation_date,start_date,expire_date,status,product_code
0,33359,2000-10-14,2000-11-07,2024-10-16,NaN,domit
1,34150,2001-12-13,2001-12-20,2026-12-19,NaN,domit
2,34249,2002-02-20,2002-02-20,2027-02-21,NaN,domcom
3,34391,2003-02-18,2003-02-19,2027-04-06,NaN,domnet
4,34706,2001-06-08,2001-06-08,2027-06-08,NaN,domit
...,...,...,...,...,...,...
170240,27546243,2023-08-08,2023-08-08,2026-05-08,SUSPENDED,deds2up_raid_h200
170241,27548220,2023-08-08,2023-08-08,2025-12-12,SUSPENDED,email25ox1q5
170242,27575634,2023-08-16,2023-08-16,2026-07-16,NaN,deds2up_hdd_3s_1000
170243,27575719,2023-08-16,2023-08-16,2026-07-16,NaN,deds2up_plesk_free


## Orders

Largest table.

The foundation for revenue.

In [74]:
orders.head()

,client_id,ordercode,renewcode,orderdate,expiredate,subscription_id,product_id,quantity,productprice,currency,duration_months,flag_recurring
0,AO9002-EURO,00000009zrna,00000009zrna,2022-03-11,2023-04-14,23395131,1839,1,0,EUR,12,1
1,AI6077-EURO,558d7f5b91a78,000000036tuu,2018-06-22,2019-06-26,9013343,359,1,26,EUR,12,1
2,GM31287-EURO,000000041ors,00000005mw9a,2020-01-14,2021-01-17,15359823,28181,0,0,EUR,12,1
3,AC84123-EURO,000000046mkc,5e2287feeacd0,2020-01-18,2021-02-07,15507842,57441,1,59,EUR,12,1
4,MG45759-EURO,00000002xlul,00000005ayzb,2019-10-24,2020-10-23,13845917,1588,0,25,EUR,12,1


In [75]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210000 entries, 0 to 209999
Data columns (total 12 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   client_id        210000 non-null  object 
 1   ordercode        210000 non-null  object 
 2   renewcode        209998 non-null  object 
 3   orderdate        210000 non-null  object 
 4   expiredate       210000 non-null  object 
 5   subscription_id  184321 non-null  float64
 6   product_id       210000 non-null  int64  
 7   quantity         210000 non-null  int64  
 8   productprice     210000 non-null  float64
 9   currency         210000 non-null  object 
 10  duration_months  210000 non-null  float64
 11  flag_recurring   210000 non-null  int64  
dtypes: float64(3), int64(3), object(6)
memory usage: 19.2+ MB


In [76]:
orders.describe().T

,count,mean,std,min,25%,50%,75%,max
subscription_id,184321,13951688,5863798,15,9982812,14659087,17818243,28235326
product_id,210000,17162,26324,5,771,1588,28882,175421
quantity,210000,1,0,0,0,1,1,1
productprice,210000,19,55,-2159,0,7,26,8174
duration_months,210000,12,4,0,12,12,12,120
flag_recurring,210000,1,0,0,1,1,1,1


### Orders - Summary

**`productprice`**
- Pct 25% = 0.01, way below median (7.45) and mean (18.94)... a quarter of rows at essentially zero... **Investigate whether the 0.01 pattern is concentrated in a specific product category or customer segment**: If it's systematic, it's not noise, it's a business artifact that needs a deliberate treatment decision before touching MRR.
- Min = -2,159 - maybe OK (likely refunds, credit adjustments)

**`subscription_id`**
- Count = 184,321 vs 210,000 total rows — ~25,679 rows have no subscription_id
- It's float64 and should be integer (in pandas this happens when a column has NULLs, which confirms the missing values above)
- **Investigate what these orders are**

**`quantity`**
- Pct 25% = 0, median = 0 — **more than half the table has zero quantity**
- Likely overlaps with the 0.01 price rows — same underlying issue? **Investigate**

**`duration_months`**
- Min = 0 — **division by zero**, risk in any MRR formula that divides price by duration
- Max = 120 (10 years) — **worth checking if real product tier or data error**
- Median = 12, annual contracts are the norm

**`flag_recurring`**
- Mean = 0.9885 - 98.85% recurring, but the 1.15% that isn't matters for MRR
- One-time charges shouldn't be spread across months as recurring revenue
- **`flag_recurring` = 0 orders must be excluded from MRR calculations** — they represent one-time charges mixed into the same table as recurring subscriptions. **Need to add this as a filter condition in the transformation layer.**

**`orderdate` / `expiredate`**
- Both read as object (string) not datetime — **need explicit casting before any date-based calculation**

### `productprice` - Is the 0.01 price pattern concentrated in a specific product category?

If it's random noise it should be spread evenly. 

If it's systematic it's a business artifact — bundled services, migration placeholders, internal line items.

In [77]:
products['product_id'] = pd.to_numeric(products['product_id'], errors='coerce')
low_price = orders[orders['productprice'] == 0.01]
low_price.merge(products[['product_id', 'product_cat_desc,,,,']], on='product_id', how='left')['product_cat_desc,,,,'].value_counts()

product_cat_desc,,,,
Applications,,,,    44362
Domains,,,,         17931
Other,,,,            4290
Hosting,,,,          3384
Name: count, dtype: int64

Pattern is present across all categories but heavily concentrated in Applications (63% of all 0.01 rows) — consistent with app activations or included add-ons where the "price" is a system placeholder, not a real charge.

#### `productprice` Action

Flag as `is_revenue_valid = FALSE`, do NOT drop — these are real order records (but generate no meaningful recurring revenue)

### `productprice` - Do the 0.01 price rows and the quantity = 0 rows overlap?

If they're the same population, they represent one single class of non-revenue records, not two separate problems.

In [78]:
overlap = ((orders['productprice'] == 0.01) & (orders['quantity'] == 0)).sum()
total_low = (orders['productprice'] == 0.01).sum()
total_zero_qty = (orders['quantity'] == 0).sum()
print(f"Both conditions: {overlap:,}")
print(f"price = 0.01 only: {total_low:,}")
print(f"quantity = 0 only: {total_zero_qty:,}")

Both conditions: 49,201
price = 0.01 only: 69,967
quantity = 0 only: 78,628


- 49,201 rows have both `productprice = 0.01` AND `quantity = 0` — these are the same population

But they're not identical:

- 20,766 rows have `price = 0.01` with `quantity > 0`
- and 29,427 have `quantity = 0` with a real price

They overlap heavily but are not the same thing — both conditions are needed independently in the `is_revenue_valid` flag, not just one of them

#### `productprice` Action

Exclude `quantity = 0` orders from MRR — `productprice × quantity = 0` regardless of price, no recurring revenue generated. 

Implemented as `quantity = 0` condition in `is_revenue_valid` flag in staging.

### Is `duration_months = 120` a real product tier or a data error?

120 months = 10 years

If it's a clean round number appearing consistently it's likely a real long-term contract tier (if it's scattered it's noise).

In [79]:
orders['duration_months'].value_counts().sort_index().tail(10)

duration_months
24     6385
25        1
36     1427
48       26
55        1
60      416
70        1
72        2
84        3
120       7
Name: count, dtype: int64

- 120 months appears only 7 times — too few to be a standard product tier
- 60 months (416 rows) looks like a real tier — significant volume
- Some odd values (25, 55, 70, 72, 84) appearing 1-3 times are likely data entry errors or edge cases
- No action needed for MRR — the productprice / duration_months formula handles all values correctly as long as duration > 0 (already guarded in staging)

#### `duration_months` Action

Flag to business: confirm whether 120-month and non-standard durations (25, 55, 70, 72, 84) are valid contract terms or data entry errors — no blocking issue for the model but worth validating

### What are the orders with no `subscription_id`?

25,679 orders have no subscription_id — before deciding how to treat them

One-time purchases? Refunds? A specific product type?

In [80]:
orders[orders['subscription_id'].isna()][['productprice', 'quantity', 'flag_recurring', 'duration_months']].describe()

,productprice,quantity,flag_recurring,duration_months
count,25679,25679,25679,25679
mean,3,1,1,12
std,30,0,0,1
min,-2159,1,0,0
25%,2,1,1,12
50%,5,1,1,12
75%,7,1,1,12
max,1085,1,1,60


- quantity = 1 for all 25,679 rows (std = 0) — no zero-quantity orders here
- flag_recurring = 1 for virtually all — these are recurring subscriptions, not one-time purchases
- productprice median = 5€, mean = 3€ — real prices, not 0.01 placeholders
- duration_months median = 12, max = 60 — normal contract lengths

Legitimate recurring orders with real revenue — Not refunds, not noise. Simply not linked to a subscription record in the subscriptions table.

#### Action:

- These orders cannot be excluded from MRR — they carry real recurring revenue
- Include them in `fct_orders` but they will have NULL `subscription_id` — they won't appear in any subscription-level analysis (churn, retention) but will contribute to MRR and ARPU correctly
- Flag for the team: 25,679 orders missing subscription linkage — source system issue, not a data quality decision we can fix in staging

### Orders - Actions

- `productprice = 0.01` is systematic, concentrated in Applications (63%) — flag as `is_revenue_valid = FALSE`, do not drop. These are real order records but generate no meaningful recurring revenue
- Keep negative prices as-is — they are refunds and must reduce MRR in the month recorded, not be corrected retroactively
- `quantity = 0` orders excluded from MRR — `productprice × quantity = 0` regardless of price. Implemented as `quantity > 0` condition in `is_revenue_valid`
- `productprice = 0.01` and `quantity = 0` overlap but are not identical (49,201 rows have both) — both conditions needed independently in `is_revenue_valid`
- Guard against `duration_months = 0` in MRR formula — treat as non-revenue row, never divide by zero. Already implemented in staging
- Non-standard `duration_months` values (25, 55, 70, 72, 84, 120) appear 1-7 times each — flag to business for validation, no blocking issue for the model
- Exclude `flag_recurring = 0` from MRR — one-time charges must not be spread across months as recurring revenue. Implemented as third condition in `is_revenue_valid`
- Cast `orderdate` and `expiredate` to datetime in staging
- Cast `subscription_id` to integer after handling nulls
- 25,679 orders with no `subscription_id` are legitimate recurring orders — include in MRR, exclude from subscription-level analysis. Flag to Data Management

## Subscriptions

In [81]:
subscriptions

,subscription_id,creation_date,start_date,expire_date,status,product_code
0,33359,2000-10-14,2000-11-07,2024-10-16,NaN,domit
1,34150,2001-12-13,2001-12-20,2026-12-19,NaN,domit
2,34249,2002-02-20,2002-02-20,2027-02-21,NaN,domcom
3,34391,2003-02-18,2003-02-19,2027-04-06,NaN,domnet
4,34706,2001-06-08,2001-06-08,2027-06-08,NaN,domit
...,...,...,...,...,...,...
170240,27546243,2023-08-08,2023-08-08,2026-05-08,SUSPENDED,deds2up_raid_h200
170241,27548220,2023-08-08,2023-08-08,2025-12-12,SUSPENDED,email25ox1q5
170242,27575634,2023-08-16,2023-08-16,2026-07-16,NaN,deds2up_hdd_3s_1000
170243,27575719,2023-08-16,2023-08-16,2026-07-16,NaN,deds2up_plesk_free


In [82]:
for col in ['creation_date', 'start_date', 'expire_date']:
    subscriptions[col] = pd.to_datetime(subscriptions[col])

subscriptions[['creation_date', 'start_date', 'expire_date']].describe()

,creation_date,start_date,expire_date
count,170245,170245,170245
mean,2017-08-25 23:28:05.187817472,2017-08-15 07:19:22.797145600,2024-05-19 22:06:22.700226304
min,1998-07-29 00:00:00,1996-12-01 00:00:00,2016-04-03 00:00:00
25%,2016-03-10 00:00:00,2016-02-16 00:00:00,2022-03-04 00:00:00
50%,2018-08-07 00:00:00,2018-08-08 00:00:00,2024-07-25 00:00:00
75%,2020-05-29 00:00:00,2020-05-29 00:00:00,2026-11-19 00:00:00
max,2023-12-13 00:00:00,2024-04-17 00:00:00,2036-06-16 00:00:00


### Subscriptions - Summary

**`status`**
- Only 48,173 non-null out of 170,245 — 72% of rows have no status value
- **The most critical finding in this table**: NULL doesn't mean unknown, it means "nothing happened yet" — **need to derive active/churned status from `expire_date` instead**

**`subscription_id`**
- Clean PK
- In orders it was float64 because of the ~25k nulls — confirms those orders exist without a matching subscription

**`creation_date` / `start_date` / `expire_date`**
- All object (string) — **same casting issue as orders, need explicit date conversion before any churn/retention calculation**
  

#### Dates:

**`creation_date` / `start_date`**
- Data goes back to 1996-1998
- Both columns move together (means and percentiles almost identical) — creation and start are essentially the same event for most subscriptions

**`expire_date`**
- Min = 2016 - some subscriptions expired years ago — historical churn already in the data
- Pct 25% = 2022 — a quarter of subscriptions expired before 2022, already churned
- Median = 2024-07-25 — more than half of all subscriptions have already expired as of today (July 2026)
- Pct 75% = 2026-11-19 — three quarters expire before end of 2026
- Max = 2036 — some very long contracts (consistent with the 120-month duration we saw in orders)
  

#### Key insight for churn definition:

The median `expire_date` being around today means the dataset captures both historical churned subscriptions and currently active ones — it's not just a snapshot of active base, it's the full lifecycle history. 

This is what makes it possible to build proper cohort retention analysis.

In [83]:
today = pd.Timestamp.today()
print((subscriptions['expire_date'] >= today).sum(), "active (expire in future)")
print((subscriptions['expire_date'] < today).sum(), "expired")

64818 active (expire in future)
105427 expired


- 64,881 subscriptions with expire_date in the future (38%) — potential active base
- 105,364 already expired (62%) — historical churn captured in the dataset
- **But `expire_date` alone is not enough to define active** — need to cross with status field, since a subscription can have a future `expire_date` but be explicitly CANCELLED or SUSPENDED
- Next: check how many of the 64,881 "future" subscriptions have a non-null problematic status

In [84]:
future = subscriptions[subscriptions['expire_date'] >= today]
future['status'].value_counts(dropna=False)

status
NaN                    62784
CANCELLED                813
TOWEPANEL                439
MAINTENANCE CHANGED      333
OWNER CHANGED            147
AWAITING PAYMENT         107
UNPAID                   105
PRE-WAIVED                83
TO VERIFY                  4
TO DELETE                  2
SUSPENDED                  1
Name: count, dtype: int64

- True active base is not 64,881 — it's closer to 62,839 (NaN status + future `expire_date`)
- CANCELLED with future `expire_date` = customer cancelled but still within paid period — not active for MRR purposes
- AWAITING PAYMENT + UNPAID = early churn signal, worth tracking separately
- This confirms the active definition needs both conditions: `expire_date >= today AND (status IS NULL OR status NOT IN ('CANCELLED', 'TO DELETE', 'SUSPENDED'))`

### Subscriptions - Actions

- Cast `creation_date`, `start_date`, `expire_date` to datetime in staging
- Derive `subscription_status_clean` field combining `expire_date` and `status` — do not use raw `status` alone
- Active definition: `expire_date >= today AND (status IS NULL OR status NOT IN ('CANCELLED', 'TO DELETE', 'SUSPENDED'))`
- Flag AWAITING PAYMENT + UNPAID separately as early churn signal — useful for a "at risk" segment
- Treat CANCELLED with future `expire_date` as churned for MRR purposes, not active

## Customers

In [85]:
customers

,client_id,client_country,join_date,segment
0,AP54723-EURO,IT,2019-01-04,Ret-E
1,AP56503-EURO,IT,2019-03-06,Ret-E
2,AP58323-EURO,IT,2019-05-17,Ret-E
3,AP58361-EURO,IT,2019-05-19,Ret-E
4,AP60022-EURO,IT,2019-07-23,Ret-E
...,...,...,...,...
92037,GC50639-EURO,IT,2017-11-14,Ret-E
92038,GC29384-EURO,IT,2012-05-08,Ret-E
92039,EN3414-EURO,IT,2019-09-27,Ret-E
92040,LM28190-EURO,IT,2018-02-15,Ret-E


In [86]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92042 entries, 0 to 92041
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   client_id       92042 non-null  object
 1   client_country  91184 non-null  object
 2   join_date       92006 non-null  object
 3   segment         92042 non-null  object
dtypes: object(4)
memory usage: 2.8+ MB


In [87]:
customers.describe().T

,count,unique,top,freq
client_id,92042,92042,AP54723-EURO,1
client_country,91184,108,IT,88388
join_date,92006,7997,2016-10-27,634
segment,92042,5,Ret-E,77813


In [88]:
customers['segment'].value_counts()

segment
Ret-E       77813
Ret-N       10624
Res          2548
obp          1021
Registry       36
Name: count, dtype: int64

In [89]:
customers['client_country'].value_counts(normalize=True).head(5)

client_country
IT   1
CH   0
GB   0
FR   0
ES   0
Name: proportion, dtype: float64

In [90]:
customers['join_date'].value_counts().head(5)

join_date
2016-10-27    634
2016-10-26    416
2020-04-02     75
2019-01-08     75
2019-01-23     74
Name: count, dtype: int64

### Customers - Summary

**`segment`**
- Ret-E (77,813) and Ret-N (10,624) dominate — together 96% of the base
- Res (2,548), obp (1,021), Registry (36) - marginal segments
- Segment labels are not documented in the case — meaning of Ret-E vs Ret-N unknown without business context, assumed to be internal customer classification

**`client_country`**
- Italian business
- Can be string but 858 NULL need to be UNKNOWN or exclude

**`join_date`**
- To datetime
- 2016-10-27 (634) and 2016-10-26 (416) — two consecutive days with volume far above anything else (next highest is 75)
- Likely a data migration event, a major promotion, a market launch...
- Worth flagging to the business before using `join_date` for cohort analysis (if it's a migration, these customers' tenure is not comparable to organically acquired ones)

### Customers - Actions

- Cast `join_date` to datetime in staging
- Map 858 null `client_country` values to 'UNKNOWN' — do not drop, country is not a join key
- Flag customers with `join_date` IN ('2016-10-26', '2016-10-27') as `is_migration_cohort` — isolate from retention metrics until business confirms the nature of that spike
- Segment labels need business validation — add a note in the data product documentation that Ret-E, Ret-N, Res, obp, Registry are assumed internal classifications with unknown meaning

## Products

In [91]:
products

,product_id,product_code,product_desc,product_subgroup_desc,product_group_desc,product_subcat_desc,"product_cat_desc,,,,"
0,1106,chgownde,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
1,1107,chgowndj,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
2,1112,chgownec,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
3,1117,chgownes,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
4,1119,chgowneu,Change Ownership,Domain services,Domain services,Domain Names,"Domains,,,,"
...,...,...,...,...,...,...,...
1457,56502,deds2up_c_xe3_1240v5,deds2up_c_xe3_1240v5,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"
1458,56503,deds2up_raid_h330,deds2up_raid_h330,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"
1459,56562,domfashion,Domain,Domain Names,Domainnames,Domain Names,"Domains,,,,"
1460,56623,deds2_dellds452,Server Dell DS 4.5.2,Dedicated Server Hosting,Dedicated Server Hosting,Professional Hosting,"Hosting,,,,"


In [92]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1462 entries, 0 to 1461
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   product_id             1461 non-null   float64
 1   product_code           1461 non-null   object 
 2   product_desc           1461 non-null   object 
 3   product_subgroup_desc  1461 non-null   object 
 4   product_group_desc     1461 non-null   object 
 5   product_subcat_desc    1461 non-null   object 
 6   product_cat_desc,,,,   1461 non-null   object 
dtypes: float64(1), object(6)
memory usage: 80.1+ KB


In [93]:
products.describe().T

,count,mean,std,min,25%,50%,75%,max
product_id,1461,41581,42160,5,1311,36303,61662,175421


In [94]:
products['product_cat_desc,,,,'].value_counts()

product_cat_desc,,,,
Domains,,,,         865
Hosting,,,,         387
Applications,,,,    147
Other,,,,            55
Hosting,,,            5
Hosting               1
Applications,,        1
Name: count, dtype: int64

### Products - Summary

**`product_id`**
- 1,462 rows, 1,462 unique — clean PK
- object to integer

**`product_code`**
- 1,461 non-null — 1 null row, minor but confirms there's one incomplete record

**`product_desc`**
- Only 605 unique values for 1,461 rows — multiple products share the same description, hierarchy lives in the category columns not in the description

**`product_cat_desc`**
- Values contain trailing commas in column name and values — this is a source data quality issue, any tool reading this file will see the same, not only pandas
- Trailing commas in values are not uniform — ranges from 1 to 4 commas per value. Fix must handle variable number.
- Fix in staging
- 7 unique categories but top value is "Domains"

**`product_subcat_desc`**
- 12 unique subcategories
- `product_subgroup_desc` has 24, `product_group_desc` has 19 — clean hierarchy exists but needs validation that it's consistent (same product always maps to the same category)

### Products - Actions

- Cast `product_id` to integer
- Strip trailing commas from `product_cat_desc` column name and values. Trailing commas in values are not uniform — ranges from 1 to 4 commas per value. Fix must handle variable number.
- Drop the 1 incomplete row (null across all descriptive fields)

## Referential integrity

We now check whether the four tables can actually be joined cleanly

Orphan IDs cause loss in the final model without warning

### **`orders` & `products`**

In [95]:
orders_products = set(orders['product_id'].unique())
products_ids = set(products['product_id'].astype(int, errors='ignore'))
print("product_ids in orders not in products:", len(orders_products - products_ids))

product_ids in orders not in products: 1


1462 is the entire products table size, which means the join is failing completely

Root cause: `product_id` in orders is integer but in products it's object (string) — the `astype(int, errors='ignore')` silently failed

In [96]:
products_ids = set(pd.to_numeric(products['product_id'], errors='coerce').dropna().astype(int).unique())
print("product_ids in orders not in products:", len(orders_products - products_ids))

product_ids in orders not in products: 1


The earlier result of 1,462 was a false positive caused by type mismatch (int vs string). 

Real orphan count is 1 (0.0005%). Keep, flag for the team.

### **`orders` & `subscriptions`**

In [97]:
orders_subs = set(orders['subscription_id'].dropna().astype(int).unique())
subs_ids = set(subscriptions['subscription_id'].unique())
print("subscription_ids in orders not in subscriptions:", len(orders_subs - subs_ids))

subscription_ids in orders not in subscriptions: 4


4 subscription_ids in orders not found in subscriptions — 0.002% but worth documenting

These orders will lose their subscription context in the join — keep them, flag for the team

### Referential Integrity - Actions

**`orders` & `products`**
- 1 `product_id` in orders has no match in products — the earlier result of 1,462 was a false positive caused by type mismatch (int vs string). Real orphan count is 1, negligible (0.0005%). Keep, flag for the team.

**`orders` & `subscriptions`**
- 4 subscription_ids in orders not found in subscriptions — 0.002% but worth documenting
- These orders will lose their subscription context in the join — keep them, flag for the team

## Actions - Global recap

### Orders - Actions

- `productprice = 0.01` is systematic, concentrated in Applications (63%) — flag as `is_revenue_valid = FALSE`, do not drop. These are real order records but generate no meaningful recurring revenue
- Keep negative prices as-is — they are refunds and must reduce MRR in the month recorded, not be corrected retroactively
- `quantity = 0` orders excluded from MRR — `productprice × quantity = 0` regardless of price. Implemented as `quantity > 0` condition in `is_revenue_valid`
- `productprice = 0.01` and `quantity = 0` overlap but are not identical (49,201 rows have both) — both conditions needed independently in `is_revenue_valid`
- Guard against `duration_months = 0` in MRR formula — treat as non-revenue row, never divide by zero. Already implemented in staging
- Non-standard `duration_months` values (25, 55, 70, 72, 84, 120) appear 1-7 times each — flag to business for validation, no blocking issue for the model
- Exclude `flag_recurring = 0` from MRR — one-time charges must not be spread across months as recurring revenue. Implemented as third condition in `is_revenue_valid`
- Cast `orderdate` and `expiredate` to datetime in staging
- Cast `subscription_id` to integer after handling nulls
- 25,679 orders with no `subscription_id` are legitimate recurring orders — include in MRR, exclude from subscription-level analysis. Flag to Data Management

### Subscriptions - Actions

- Cast `creation_date`, `start_date`, `expire_date` to datetime in staging
- Derive `subscription_status_clean` field combining `expire_date` and `status` — do not use raw `status` alone
- Active definition: `expire_date >= today AND (status IS NULL OR status NOT IN ('CANCELLED', 'TO DELETE', 'SUSPENDED'))`
- Flag AWAITING PAYMENT + UNPAID separately as early churn signal — useful for a "at risk" segment
- Treat CANCELLED with future `expire_date` as churned for MRR purposes, not active

### Customers - Actions

- Cast `join_date` to datetime in staging
- Map 858 null `client_country` values to 'UNKNOWN' — do not drop, country is not a join key
- Flag customers with `join_date` IN ('2016-10-26', '2016-10-27') as `is_migration_cohort` — isolate from retention metrics until business confirms the nature of that spike
- Segment labels need business validation — add a note in the data product documentation that Ret-E, Ret-N, Res, obp, Registry are assumed internal classifications with unknown meaning

### Products - Actions

- Cast `product_id` to integer
- Strip trailing commas from `product_cat_desc` column name and values. Trailing commas in values are not uniform — ranges from 1 to 4 commas per value. Fix must handle variable number.
- Drop the 1 incomplete row (null across all descriptive fields)